### PR TITLE
Contract MPT storage and Remote State Fetching Support

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

This PR contains these following points:
1. Support Remote State Fetching
2. Contract Merkle Partricia Tree storage
3. Data migrate
Since these two features are dependent to each others so submitting in one PR as a whole.
### Support Remote State Fetching
We mainly introduced:
- Constant `TYPE_INDICATOR` for all scilla data entry, this is introduced by parsing the result from Scilla Checker
- Constant `CONTRACT_ADDR_INDICATOR` for quick locating a remote contract.
- Constant `SCILLA_VERSION_INDICATOR` for version consistency when reading states from remote contract.
- Function `ContractStorage2::FetchExternalStateValue` algorithm for fetching external contract state value.
- Some refactoring to related functions
### Contract MPT storage
We mainly introduced:
- A series of new classes in `ContractStorage2Data.h` (e.g. `AddDeleteMap`, `RecordableAddDeleteMap` ...) for interacting with Trie library.
- Adopted new composing mechanism for contract storage root based on MPT.
- Revised most of contract storage related basic functions for supporting MPT storage.
### Data migrate
- Basically we will let every contracts existing in the previous mainnet to run through the Scilla checker again to make up the new parameters needed and deprecate the old ones.
 
## Backward Compatibility

- [ ] This is not a breaking change
- [x] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libPersistence/ContractStorage2.cpp [hunk 1]
```diff
@@ -1021,46 +1127,89 @@ void ContractStorage2::InitTempState(bool callFromExternal) {
   }
 }
 
-dev::h256 ContractStorage2::GetContractStateHashCore(const dev::h160& address,
-                                                     bool temp) {
-  if (IsNullAddress(address)) {
-    LOG_GENERAL(WARNING, "Null address rejected");
-    return dev::h256();
-  }
+dev::h256 ContractStorage2::UpdateContractTrieCore(
```
Comment (ansnunez) [inline src/libPersistence/ContractStorage2.cpp:1130]: Just to confirm: we have removed direct data hashing and will use trie hashing regardless of state entries count?

File: src/libPersistence/ContractStorage2.cpp [hunk 2]
```diff
@@ -342,17 +393,57 @@ bool ContractStorage2::FetchStateValue(const dev::h160& addr, const bytes& src,
...
+  // get caller version if not available
+  if (caller_version == std::numeric_limits<uint32_t>::max()) {
+    std::map<std::string, bytes> t_caller_version;
+    string version_key =
+        GenerateStorageKey(caller, SCILLA_VERSION_INDICATOR, {});
+    FetchStateDataForKey(t_caller_version, version_key, true);
+    if (t_caller_version.empty()) {
+      return false;
+    }
+    caller_version = std::stoul(
```
Comment (ansnunez) [inline src/libPersistence/ContractStorage2.cpp:409]: `stoul` may throw exception

File: src/libPersistence/ContractStorage2.cpp [hunk 3]
```diff
@@ -342,17 +393,57 @@ bool ContractStorage2::FetchStateValue(const dev::h160& addr, const bytes& src,
...
+  }
+
+  // get target version if not available
+  std::map<std::string, bytes> t_target_version;
+  string version_key = GenerateStorageKey(target, SCILLA_VERSION_INDICATOR, {});
+  FetchStateDataForKey(t_target_version, version_key, true);
+  if (t_target_version.empty()) {
+    return false;
+  }
+  uint32_t target_version = std::stoul(
```
Comment (ansnunez) [inline src/libPersistence/ContractStorage2.cpp:420]: `stoul` may throw exception

**Review by ansnunez — CHANGES_REQUESTED** [review]

File: src/libPersistence/ContractStorage2.cpp [hunk 4]
```diff
@@ -878,65 +934,89 @@ bool ContractStorage2::UpdateStateValue(const dev::h160& addr, const bytes& q,
 void ContractStorage2::UpdateStateDatasAndToDeletes(
     const dev::h160& addr, const std::map<std::string, bytes>& t_states,
     const std::vector<std::string>& toDeleteIndices, dev::h256& stateHash,
-    bool temp, bool revertible) {
+    bool temp, bool revertible, bool updateRoot) {
```
[the remark this answers is not in the document]
Comment (kaikawaliu) [reply src/libPersistence/ContractStorage2.cpp:937]: great finding! the `false` case was omitted during the development, now we can remove it

File: src/libPersistence/ContractStorage2.cpp (hunk 1 above)
Comment (kaikawaliu) [reply src/libPersistence/ContractStorage2.cpp:1130]: No we will still direct hash the states for contracts with map, which can be identified by fetching the metadata parsed for remote state fetching. Thanks for reminding, this part of logic however has not been implemented yet, will do ASAP.

**Review by ansnunez — APPROVED** [review]

**Review by sandipbhoir — APPROVED** [review]
